### PR TITLE
Context-aware Menus

### PR DESCRIPTION
**Pull request by geralddejong**
Menus only appear when they apply
TinkererAction nested in CrucibleAction

### CODE REVIEW

**Comment by mitchelljustin** [comment]
I think you meant to open this PR against `develop`, so I changed it for you

File: src/user_interface/menu.rs [hunk 1]
```diff
@@ -192,4 +251,6 @@ fn to_key_code(ch: char) -> Option<VirtualKeyCode> {
         'Z' => Z,
         _ => return None
     })
-}
\ No newline at end of file
+}
+
+const ALWAYS: fn(MenuEnvironment) -> bool = |_| true;
```
Comment (mitchelljustin) [inline src/user_interface/menu.rs:256]: Is there a reason this isn't just a function?

File: src/user_interface/menu.rs [hunk 2]
```diff
@@ -9,41 +8,86 @@ use crate::crucible::CrucibleAction;
...
+pub struct MaybeMenu {
+    exists_in: fn(MenuEnvironment) -> bool,
+    menu: Menu,
+}
+
+impl MaybeMenu {
+    pub fn menu_in(&self, environment: MenuEnvironment) -> Option<Menu> {
+        (self.exists_in)(environment).then_some(self.menu.clone())
+    }
+}
```
Comment (mitchelljustin) [inline src/user_interface/menu.rs:24]: Cool way of doing this

File: src/user_interface/menu.rs [hunk 3]
```diff
@@ -72,96 +116,111 @@ impl Menu {
...
-                Menu::action(label, Action::Crucible(CrucibleAction::SetSpeed(speed))))
-            .collect()
+    fn fabric_menu(fabrics: &[FabricPlan]) -> Menu {
+        Self::fabric_menu_recurse(Menu::new("Fabrics"), fabrics, Vec::new())
+    }
+
+    fn speed_menu() -> Menu {
+        let mut menu = Menu::new("Speed");
+        for (speed, label) in [(0usize, "Paused"), (5, "Glacial"), (25, "Slow"), (125, "Normal"), (625, "Fast")] {
+            menu = menu.action(label, true, |_| true, Action::Crucible(CrucibleAction::SetSpeed(speed)));
```
Comment (mitchelljustin) [inline src/user_interface/menu.rs:146]:
```suggestion
            menu = menu.action(label, true, ALWAYS, Action::Crucible(CrucibleAction::SetSpeed(speed)));
```
Unless I misunderstand

**Review by mitchelljustin — APPROVED** [review]
You're building a bona fide UI framework here, impressive.

File: src/user_interface/menu.rs (hunk 1 above)
Comment (geralddejong) [reply src/user_interface/menu.rs:256]: I consider it to be a constant.

**Pull request merged**